### PR TITLE
Admin context listener integration

### PR DESCRIPTION
**Pull request by thysw95**
Added integration between the admin context module and the listener, so that the listener would check for privileges, before executing commands

## Summary by Sourcery

Integrate the admin context module with the listener to enforce privilege checks before command execution. Enhance the listener by adding new data classes for session and context management, and update the module retrieval process to include user identity information.

New Features:
- Integrate privilege checking into the command execution process, ensuring that users have the necessary permissions before executing commands.

Enhancements:
- Extend the `get_marketplace_module_by_name` function to include user identity in the request, improving module retrieval accuracy.
- Add new data classes `sessionData` and `contextData` to manage session and context information more effectively.

### CODE REVIEW

File: modules/WaddleBot-Listener/src/listener/listener.py [hunk 1]
```diff
@@ -635,8 +659,54 @@ def get_command_properties(self, commandlist: list, metadata: dict) -> commandDa
             else:
                 return None
                 
-            
+        return None
+    
+    # Function to return a flag depending on whether the given admin command can be excuted or not. 
+    # For this check to work, it receives the session data of the user, the context of the user, and 
+    # the command data of the command to be executed.
+    def check_command_priv_list(self, sessionData: sessionData, contextData: contextData, commandData: commandData, user_priv_list: list) -> bool:
```
[the remark this answers is not in the document]
Comment (thysw95) [reply modules/WaddleBot-Listener/src/listener/listener.py:667]: I understand this is crucial to security. I did test this function beforehand with multiple client and it looks like it correctly checks the privileges

[the remark this answers is not in the document]
Comment (thysw95) [reply modules/WaddleBot-Listener/src/listener/listener.py:667]: Ok got ya! Ill discuss with the other developers when we work through this part.

File: modules/WaddleBot-Listener/src/listener/listener.py [hunk 2]
```diff
@@ -635,8 +659,54 @@ def get_command_properties(self, commandlist: list, metadata: dict) -> commandDa
...
+        logging.info("Checking Admin Command....")
+
+        req_priv_list = commandData['req_priv_list']
+
+        # Check if the user has the required priv_list to execute the command
+        if len(req_priv_list) > 0 and len(user_priv_list) > 0:
+            if not all(item in user_priv_list for item in req_priv_list):
+                return False
```
[the remark this answers is not in the document]
Comment (thysw95) [reply modules/WaddleBot-Listener/src/listener/listener.py]: Hmmmm
This particular issue was resolved in the previous suggestion that completely rewrote the functions. Ensure that you try to organize most issues into one suggestion. I know if statement nesting is a problem, so ill try to pay more attention next time

File: modules/WaddleBot-Listener/src/listener/listener.py [hunk 3]
```diff
@@ -635,8 +659,54 @@ def get_command_properties(self, commandlist: list, metadata: dict) -> commandDa
...
                 return None
                 
-            
+        return None
+    
+    # Function to return a flag depending on whether the given admin command can be excuted or not. 
+    # For this check to work, it receives the session data of the user, the context of the user, and 
+    # the command data of the command to be executed.
+    def check_command_priv_list(self, sessionData: sessionData, contextData: contextData, commandData: commandData, user_priv_list: list) -> bool:
+        logging.info("Checking Admin Command....")
```
[the remark this answers is not in the document]
Comment (thysw95) [reply modules/WaddleBot-Listener/src/listener/listener.py]: Is an outdated suggestion

File: modules/WaddleBot-Listener/src/listener/listener.py [hunk 4]
```diff
@@ -596,15 +612,23 @@ def display_help(self) -> str:
         return helpMessage
     
     # Function to retrieve a marketplace module entry by its URL
-    def get_marketplace_module_by_name(self, moduleName: str) -> marketplaceModuleData:
+    def get_marketplace_module_by_name(self, moduleName: str, identity_name: str) -> marketplaceModuleData:
         logging.info("Getting Marketplace Module by URL....")
 
         try:
+            # URL encode the module name
+            moduleName = quote(moduleName)
```
[the remark this answers is not in the document]
Comment (thysw95) [reply modules/WaddleBot-Listener/src/listener/listener.py:620]: Ill keep it in mind

File: modules/WaddleBot-Listener/src/listener/listener.py [hunk 5]
```diff
@@ -635,8 +659,54 @@ def get_command_properties(self, commandlist: list, metadata: dict) -> commandDa
...
+        sessionExpiry = sessionData['session_expires']
 
+        # Convert the session expiry date string to a datetime object
+        sessionExpiry = time.mktime(time.strptime(sessionExpiry, "%Y-%m-%d %H:%M:%S"))
+
+        # Check if the session expiry time is greater than the current time
+        if sessionExpiry > currentTime:
+            return True
+        else:
+            return False
```
[the remark this answers is not in the document]
Comment (thysw95) [reply modules/WaddleBot-Listener/src/listener/listener.py]: This is an outdated issue that has been fixed in a previous suggestion

**Review by PenguinzPlays — APPROVED** [review]

**Pull request merged**